zio-http/jvm/src/main/scala/zio/http/netty/client/NettyConnectionPool.scala

```diff
@@ -106,38 +108,136 @@ private[netty] object NettyConnectionPool {
 
     for {
       resolvedHosts <- dnsResolver.resolve(location.host)
-      hosts         <- Random.shuffle(resolvedHosts.toList)
-      hostsNec      <- ZIO.succeed(NonEmptyChunk.fromIterable(hosts.head, hosts.tail))
-      ch            <- collectFirstSuccess(hostsNec) { host =>
-        ZIO.suspend {
-          val bootstrap = new Bootstrap()
-            .channelFactory(channelFactory)
-            .group(eventLoopGroup)
-            .remoteAddress(new InetSocketAddress(host, location.port))
-            .withOption[Integer](ChannelOption.CONNECT_TIMEOUT_MILLIS, connectionTimeout.map(_.toMillis.toInt))
-            .handler(initializer)
-          localAddress.foreach(bootstrap.localAddress)
-
-          val channelFuture = bootstrap.connect()
-          val ch            = channelFuture.channel()
-          Scope.addFinalizer {
-            NettyFutureExecutor.executed {
-              channelFuture.cancel(true)
-              ch.close()
-            }.when(ch.isOpen).ignoreLogged
-          } *> NettyFutureExecutor.executed(channelFuture).as(ch)
-        }
-      }
+      ch            <-
+        // Use Happy Eyeballs algorithm
+        happyEyeballsConnect(
+          resolvedHosts,
+          channelFactory,
+          eventLoopGroup,
+          location,
+          initializer,
+          connectionTimeout,
+          localAddress,
+        )
     } yield ch
   }
 
-  private def collectFirstSuccess[R, E, A, B](
-    as: NonEmptyChunk[A],
-  )(f: A => ZIO[R, E, B])(implicit trace: Trace): ZIO[R, E, B] = {
-    ZIO.suspendSucceed {
-      val it                 = as.iterator
-      def loop: ZIO[R, E, B] = f(it.next()).catchAll(e => if (it.hasNext) loop else ZIO.fail(e))
-      loop
+  /**
+   * Attempts to connect to a single address.
+   */
+  private def connectToAddress(
+    host: InetAddress,
+    channelFactory: JChannelFactory[JChannel],
+    eventLoopGroup: JEventLoopGroup,
+    location: URL.Location.Absolute,
+    initializer: ChannelInitializer[JChannel],
+    connectionTimeout: Option[Duration],
+    localAddress: Option[InetSocketAddress],
+  )(implicit trace: Trace): ZIO[Scope, Throwable, JChannel] = {
+    ZIO.suspend {
+      val bootstrap = new Bootstrap()
+        .channelFactory(channelFactory)
+        .group(eventLoopGroup)
+        .remoteAddress(new InetSocketAddress(host, location.port))
+        .withOption[Integer](ChannelOption.CONNECT_TIMEOUT_MILLIS, connectionTimeout.map(_.toMillis.toInt))
+        .handler(initializer)
+      localAddress.foreach(bootstrap.localAddress)
+
+      val channelFuture = bootstrap.connect()
+      val ch            = channelFuture.channel()
+      Scope.addFinalizer {
+        NettyFutureExecutor.executed {
+          channelFuture.cancel(true)
+          ch.close()
+        }.when(ch.isOpen).ignoreLogged
+      } *> NettyFutureExecutor.executed(channelFuture).as(ch)
```

Comment:
I believe this should run uninterruptibly 

```suggestion
      }.uninterruptible *> NettyFutureExecutor.executed(channelFuture).as(ch)
```